src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/ProjectReloadManager.cs

```diff
@@ -0,0 +1,508 @@
+        /// </summary>
+        private async Task ConnectToSolutionEvents()
+        {
+            await _threadHandling.SwitchToUIThread();
+
+            if (_solutionEventsCookie == VSConstants.VSCOOKIE_NIL)
+            {
+                IVsSolution solution = _serviceProvider.GetService<IVsSolution, SVsSolution>();
+                if (solution != null)
+                {
+                    int hr = solution.AdviseSolutionEvents(this, out _solutionEventsCookie);
+                    System.Diagnostics.Debug.Assert(ErrorHandler.Succeeded(hr) && _solutionEventsCookie != VSConstants.VSCOOKIE_NIL);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Removes the file change watch on the project file. 
+        /// </summary>
+        private async Task DisconnectFromSolutionEvents()
+        {
+            if(_solutionEventsCookie != VSConstants.VSCOOKIE_NIL)
+            {
+                await _threadHandling.SwitchToUIThread();
+
+                IVsSolution solution = _serviceProvider.GetService<IVsSolution, SVsSolution>();
+                if (solution != null)
+                {
+                    int hr = solution.UnadviseSolutionEvents(_solutionEventsCookie);
+                    System.Diagnostics.Debug.Assert(ErrorHandler.Succeeded(hr));
+                    _solutionEventsCookie = VSConstants.VSCOOKIE_NIL;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Handles one time initialization
+        /// </summary>
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            _reloadDelayScheduler = new TaskDelayScheduler(TimeSpan.FromMilliseconds(ReloadDelay), _threadHandling, CancellationToken.None);
+            await ConnectToSolutionEvents().ConfigureAwait(false);
+
+        }
+
+        /// <summary>
+        /// IDisposable handler. Should only be called once
+        /// </summary>
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            if(_reloadDelayScheduler != null)
+            {
+                _reloadDelayScheduler.Dispose();
+                _reloadDelayScheduler = null;
+            }
+            await DisconnectFromSolutionEvents().ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Called when one of the project files changes. In our case since only one file is watched with each cookie so the list of files
+        /// should be one.
+        /// </summary>
+        public int FilesChanged(uint cChanges, string[] rgpszFile, uint[] grfChange)
+        {
+            if(cChanges == 1 && (grfChange[0] & (uint)(_VSFILECHANGEFLAGS.VSFILECHG_Size | _VSFILECHANGEFLAGS.VSFILECHG_Time)) != 0)
+            {
+                lock(_changedProjects)
+                {
+                    var changedProject = _registeredProjects.FirstOrDefault(kv => kv.Key.ProjectFile.Equals(rgpszFile[0], StringComparison.OrdinalIgnoreCase)).Key;
```

Comment:
Also, should that be a `?.Key`, just in case the `FirstOrDefault` call returns null?